src/daemon.rs

```diff
@@ -445,6 +454,79 @@ impl DaemonContext {
         }
         Ok(())
     }
+
+    fn dht_key_to_fps(&self, key: &str) -> Vec<Fingerprint> {
+        let mut all_nodes: Vec<Fingerprint> = self.relay_graph.read().all_nodes().collect();
+        all_nodes.sort_unstable_by_key(|fp| *blake3::hash(&(key, fp).stdcode()).as_bytes());
+        all_nodes
+    }
+
+    pub async fn dht_insert(&self, key: Fingerprint, value: HavenLocator) {
+        let replicas = self.dht_key_to_fps(&key.to_string());
+
+        for replica in replicas.into_iter().take(DHT_REDUNDANCY) {
+            log::debug!("key {key} inserting into remote replica {replica}");
+            let gclient = GlobalRpcClient(GlobalRpcTransport::new(self.clone(), replica));
+            match gclient
+                .dht_insert(key, value.clone(), false)
+                .timeout(Duration::from_secs(10))
+                .await
+            {
+                Some(Err(e)) => log::debug!("inserting {key} into {replica} failed: {:?}", e),
+                None => log::debug!("inserting {key} into {replica} timed out"),
+                _ => {}
+            }
+        }
+    }
+
+    pub async fn dht_get(&self, key: Fingerprint) -> Option<HavenLocator> {
+        let replicas = self.dht_key_to_fps(&key.to_string());
+        let mut gatherer = FuturesUnordered::new();
+        for replica in replicas.into_iter().take(DHT_REDUNDANCY) {
+            gatherer.push(async move {
+                let gclient = GlobalRpcClient(GlobalRpcTransport::new(self.clone(), replica));
+                anyhow::Ok(
+                    gclient
+                        .dht_get(key, false)
+                        .timeout(Duration::from_secs(30))
+                        .await
+                        .context("timed out")??,
+                )
+            })
+        }
+        while let Some(result) = gatherer.next().await {
+            match result {
+                Err(err) => log::warn!("error while dht_get: {:?}", err),
+                Ok(None) => continue,
+                Ok(Some(val)) => return Some(val),
+            }
+        }
+        None
+    }
+
+    pub async fn insert_rendezvous(&self, locator: HavenLocator) -> Result<(), DhtError> {
+        let id_pk = locator.id_pk();
+        let payload = locator.signable().stdcode();
+        id_pk.verify(&payload, &locator.signature())?;
```

Comment:
Verification here is not useful unless as a bug-catching sanity check. This code runs on the client side, where we've already signed the data. We should instead by verifying on the server side.